### PR TITLE
Sync requires_grad after _set_tensor_requires_grad under torch.func

### LINKED ISSUE #193279: Dynamo silently drops custom `backward` of `autograd.Function` under `torch.func.grad`, producing wrong gradients
## 🐛 Describe the bug

When `torch.func.grad` is applied to a `torch.autograd.Function` inside a `torch.compile`d function, Dynamo inlines the `forward()` as plain code and **silently discards the custom `backward` staticmethod**. The functorch Grad interpreter then differentiates the inlined primitives with their default rules, so the compiled function returns a *different* gradient than eager, with no error, no warning and no graph break, even under `fullgraph=True`.

This is the reverse-mode counterpart of #192985 and shares its root cause.

### Minimal repro

The custom `backward` deliberately returns zeros, not the true derivative of `sin`, making the drop directly observable:

```python
import torch


class SinWithZeroBackward(torch.autograd.Function):
    @staticmethod
    def forward(x):
        return torch.sin(x)

    @staticmethod
    def setup_context(ctx, inputs, output):
        ctx.save_for_backward(*inputs)

    @staticmethod
    def backward(ctx, grad):
        return torch.zeros_like(grad)


def loss(x):
    return SinWithZeroBackward.apply(x).sum()


x = torch.randn(4)

print(torch.func.grad(loss)(x))  # zeros: custom backward respected
print(torch.compile(torch.func.grad(loss), backend="aot_eager", fullgraph=True)(x))
print(torch.cos(x))  # what the compiled version actually returns

# Reverse mode outside torch.func is handled correctly, for comparison:
xg = x.clone().requires_grad_()
torch.compile(loss, backend="aot_eager")(xg).backward()
print(xg.grad)  # zeros: custom backward respected
```

Output:

```
tensor([0., 0., 0., 0.])
tensor([0.9517, 0.9837, 0.4671, 0.8350])
tensor([0.9517, 0.9837, 0.4671, 0.8350])
tensor([0., 0., 0., 0.])
```

The second line is the compiled gradient and the third is `cos(x)`. They are identical, so the custom `backward` was replaced by the default derivative of the inlined `sin`.

Reproduces with `backend="eager"` and `backend="aot_eager"`, with and without `fullgraph=True`. `torch.func.vjp` is affected the same way. Reverse mode outside `torch.func`, via `.backward()`, is handled correctly.

### Potential cause

The same gate as #192985. `AutogradFunctionVariable.call_apply` routes to the `AutogradFunctionApply` higher-order op, which is what preserves the custom `backward`, only under:

```python
if requires_grad and torch.is_grad_enabled():
```

https://github.com/pytorch/pytorch/blob/af55402990611787168cfabfb457c9d30f5a4d73/torch/_dynamo/variables/misc.py#L1113

Tensors wrapped by a functorch Grad interpreter report `requires_grad=False`, so the gate evaluates falsy and execution falls through to the plain-inline branch at the bottom of `call_apply`, which drops all autograd customization.

#134784 raised this general `requires_grad` concern for `autograd.Function` under functorch transforms back in 2024, and the `autograd.Function` interaction was left open there. This is a concrete silent-wrong-result case for it.

It is also worth noting that the existing test `test_new_style_autograd_function_with_grad_compiled` in `test/dynamo/test_autograd_function.py` cannot catch this. Its Function uses `forward = x * 2` with `backward = grad * 2`, which is the true derivative, so a dropped custom `backward` is indistinguishable from an honored one.

### Expected behavior

Either of:

1. Dynamo graph breaks when a functorch Grad or Vjp transform is active, so the eager `autograd.Function` machinery, including the custom `backward`, is used; or
2. the custom `backward` is captured and used in the compiled graph.

Silently computing a different gradient than eager should not be possible.

### Versions

```
PyTorch version: 2.15.0.dev20260812+cpu
Is debug build: False
OS: Microsoft Windows 11 Home (10.0.26200 64-bit)
Python version: 3.12.10 (tags/v3.12.10:0cc8128, Apr  8 2025, 12:21:36) [MSC v.1943 64 bit (AMD64)] (64-bit runtime)
Is CUDA available: False
[pip3] numpy==2.5.2
[pip3] torch==2.15.0.dev20260812+cpu
```

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @bdhirsh @bobrenjc93 @aorenste @zou3519 @Chillee @samdow @kshitij12345 @yanboliang

### PR DESCRIPTION
**Pull request by robertomeroni**
## Issue

Fixes #193279

## Summary

`_create_differentiable` flips `requires_grad` via `_set_tensor_requires_grad`, an in-graph function whose mutation never reached its `TensorVariable`. `call_apply` gates autograd customization on that stale `False` (`misc.py:1118`), so it inlined `forward` and dropped the custom `backward`. The fix re-syncs metadata after fake propagation; the graph node is unchanged. Any `x.requires_grad` read in a transform was stale too - second test.

Behaviour change: a custom `vjp` or `jvp` now hits its pre-existing graph break under `func.grad` (`misc.py:1143`, `:1158`); both were previously inlined, so a `fullgraph=True` compile that happened to be correct now errors, as `.backward()` already does. It also fails #193277's new `test_custom_jvp_under_func_grad_does_not_graph_break`; whichever lands second updates it. Happy to rebase behind it (same reporter, forward-mode half).

## Checklist

- [x] Passes lint (`spin fixlint`)
- [x] Added/updated tests
- [ ] Updated documentation (if applicable)
- [ ] Included benchmark results (for PRs impacting perf)

## BC-breaking?

No

---
AI assistance was used (Opus 5)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Comment by robertomeroni** [comment]
the red checks here are all from the GitHub incident: every failed job ends at `Failed to download archive https://codeload.github.com/... after 3 attempts`

**Comment by robertomeroni** [comment]
@Skylion007 could you run `@pytorchbot rebase` here when you get a chance? The 86 red checks are all from the GitHub Actions incident
